Change Drivetrain Deadband

lesereli000 opened the pull request:
Adds a couple of changes to the deadband on drivetrain:
<ul>
<li>add a lowered power cap</li>
<li>add a "boost" button to raise power cap</li>
<li>change algorithm for determining power to be non-linear</li>
<li>add a translation to the curve so that when the deadband is no longer active there is no jump in power</li>
</ul>

Commit:
add requested changes to deadband system

Commit:
fix expressions for return

Commit:
fix signs for negative power

Commit:
clean up code to include ?: operators

Commit:
fix compile error and documentation

Commit:
fix signs on refactor commit

sreeves750 committed:
Change deadband command to use shorthand

tricktrap commented on src/main/java/frc/robot/Drivetrain.java:
```diff
@@ -28,8 +28,8 @@
      */
-    public void driveLeft(double speed){
-        double sp = deadband(speed);
+    public void driveLeft(double speed, boolean boost){
```
I like the idea of boost, but I think we should move the implementation "up" to avoid making autonomous code have to deal with it. I recommend moving it outside of the class entirely, and only using it in the teleopPeriodic functions.

lesereli000 replied:
boost is now called from tank drive and arcade drive methods. 

public void tankDrive(double leftDrive, double rightDrive, boolean boost){
        this.driveLeft(boost(leftDrive, boost));
        this.driveRight(boost(rightDrive, boost));
    }

tricktrap commented on src/main/java/frc/robot/Drivetrain.java:
```diff
@@ -75,9 +75,11 @@ public void tankDrive(double leftDrive, double rightDrive){
      */
-    public double deadband(double in) {
-        return Math.abs(in) > DEADBAND_LIMIT ? in : 0.0;
+    public double deadband(double in, boolean boost) {
```
This function is actually doing a little too much at this point. It's doing both "deadband" and "boost" functions which should probably be separated into their own functions. Once you do so, `deadband` doesn't need to be in the drivetrain class at all, in fact it might be useful to keep separate so that other classes can use it as well (manual control of the shooter angle, collection system, etc.) 

The boost code itself is pretty hard to follow, I think we should move all instances of `0.6` to a constant (also to make it easy to change later). Doubling up the terenary operators compacts the code but makes it harder to read and debug, it's worth having extra whitespace to make this easier to read.

lesereli000 replied:
methods are separated and deadband is in robot class

tricktrap commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -160,7 +160,7 @@ public void teleopPeriodic() {
       }
 
-      drive.arcadeDrive(turnInput, speedInput);
+      drive.arcadeDrive(turnInput, speedInput, driver.getAButtonPressed());
```
I think this is the best place to call the boost code (hopefully as a function). Rather than pass whether to boost down two or more layers of function calls, we can just adjust turnInput and speedInput here instead.

lesereli000 replied:
implemented

tricktrap reviewed (changes requested):
Ok, so the intent and final result of this code is good (it will almost certainly work), but I think it should be organized a little differently to be more maintainable.

Commit:
fix requested changes

sreeves750 committed:
Merge branch 'master' into drivetrain-deadband

tricktrap committed:
Add simple test of deadband

tricktrap committed:
Remove unnecessary fake test file

tricktrap committed:
Add more tests of deadband

tricktrap committed:
Add commented-out test to indicate that we still need work on the deadband function

tricktrap committed:
Refactor deadband into own class

tricktrap committed:
Finish refactor

tricktrap committed:
implement boost mode as an InputScaler

tricktrap reviewed (approved):
Even though I totally hijacked this PR I still approve

sreeves750 reviewed (approved):
Looks good. All that's left is to throw it on the drivetrain and make sure it works.

sreeves750 committed:
Merge branch 'master' into drivetrain-deadband

sreeves750 committed:
Merge branch 'master' into drivetrain-deadband

dracco1993 committed:
Merge branch 'master' into drivetrain-deadband

dracco1993 committed:
Fixed bad merge

Commit:
Fix Jordan's mess
I did have to remove the SmartDashboard LightClear to get it to compile.
I do not know what is the equivlant for PhotoswitchSensor.

TheBitEffect reviewed (approved):
It compiles! \\(^◇^)/

sreeves750 committed:
Update id numbers

rar1741programmer committed:
Tested deadbanding

rar1741programmer committed:
Fix PTO things
